tests/core/middleware/test_transaction_signing.py

```diff
@@ -0,0 +1,304 @@
+import pytest
+
+import eth_account
+import eth_keys
+from eth_utils import (
+    to_bytes,
+    to_hex,
+)
+from hexbytes import (
+    HexBytes,
+)
+
+from web3 import Web3
+from web3.exceptions import (
+    InvalidAddress,
+)
+from web3.middleware import (
+    construct_result_generator_middleware,
+    construct_sign_and_send_raw_middleware,
+)
+from web3.middleware.signing import (
+    gen_normalized_accounts,
+)
+from web3.providers import (
+    BaseProvider,
+)
+from web3.providers.eth_tester import (
+    EthereumTesterProvider,
+)
+from web3.utils.toolz import (
+    merge,
+    valfilter,
+)
+
+PRIVATE_KEY_1 = to_bytes(
+    hexstr='0x6a8b4de52b288e111c14e1c4b868bc125d325d40331d86d875a3467dd44bf829')
+
+ADDRESS_1 = '0x634743b15C948820069a43f6B361D03EfbBBE5a8'
+
+PRIVATE_KEY_2 = to_bytes(
+    hexstr='0xbf963e13b164c2100795f53e5590010f76b7a91b5a78de8e2b97239c8cfca8e8')
+
+ADDRESS_2 = '0x91eD14b5956DBcc1310E65DC4d7E82f02B95BA46'
+
+KEY_FUNCS = (
+    eth_keys.keys.PrivateKey,
+    eth_account.Account.privateKeyToAccount,
+    HexBytes,
+    to_hex,
+    lambda bytes_key: bytes_key,
+)
+
+
+SAME_KEY_MIXED_TYPE = tuple(key_func(PRIVATE_KEY_1) for key_func in KEY_FUNCS)
+
+MIXED_KEY_MIXED_TYPE = tuple(
+    key_func(key) for key in [PRIVATE_KEY_1, PRIVATE_KEY_2] for key_func in KEY_FUNCS
+)
+
+SAME_KEY_SAME_TYPE = (
+    eth_keys.keys.PrivateKey(PRIVATE_KEY_1),
+    eth_keys.keys.PrivateKey(PRIVATE_KEY_1)
+)
+
+MIXED_KEY_SAME_TYPE = (
+    eth_keys.keys.PrivateKey(PRIVATE_KEY_1), eth_keys.keys.PrivateKey(PRIVATE_KEY_2)
+)
+
+
+class DummyProvider(BaseProvider):
+    def make_request(self, method, params):
+        raise NotImplementedError("Cannot make request for {0}:{1}".format(
+            method,
+            params,
+        ))
+
+
+@pytest.fixture()
+def result_generator_middleware():
+    return construct_result_generator_middleware({
+        'eth_sendRawTransaction': lambda *args: args,
+        'net_version': lambda *_: 1,
+    })
+
+
+@pytest.fixture()
+def w3_base():
+    return Web3(providers=[DummyProvider()], middlewares=[])
+
+
+@pytest.fixture()
+def w3_dummy(w3_base, result_generator_middleware):
+    w3_base.middleware_stack.add(result_generator_middleware)
+    return w3_base
+
+
+@pytest.mark.parametrize(
+    'method,strict_account_matching,key_object,from_,expected',
+    (
+        ('eth_sendTransaction', True, SAME_KEY_MIXED_TYPE, ADDRESS_1, 'eth_sendRawTransaction'),
+        ('eth_sendTransaction', True, SAME_KEY_MIXED_TYPE, ADDRESS_2, ValueError),
+        ('eth_sendTransaction', False, SAME_KEY_MIXED_TYPE, ADDRESS_2, NotImplementedError),
```

Comment:
There are other places in pytests where we do this.  I agree it would be clearer to define our own exception.  I will make a note of this, and add it to another PR that includes the other tests as well.